SI-7915 Corrected range positions created during default args expansion

dotta opened the pull request:
The tree created during expansion of default arguments contained trees with the
wrong type of positions. In particular:
- The synthetic $qual should have a transparent position.
- The new Select tree should always have a range position, because `selected`
  (i.e., the called method) is always visible in the source.
- The Block that contains the expanded default arguments is transparent.

The only tricky part of the fix is the position assigned to the new Select
tree, which needs to respect the range position's invariants. In the specific
case, we ought to make sure that range positions don't overlap. Therefore, the
position assigned to the new Select tree is computed by intersecting the
original Select position (i.e., `baseFun`'s position) and the original
qualifier's position (i.e., `qual`'s position). Furthermore, calling
`withStart` would turn any position in a range position, which isn't
desiderable in general (and, even worse, this could lead to generation of
invalid - bad offsets - positions). Hence, the computation is only performed
when both `baseFun` and `qual` positions are range positions. I expect this to
be always the case if the compiler is started with -Yrangepos. In fact, I
wonder if I should add an assertion that validates this assumption.

dotta committed:
SI-7915 Corrected range positions created during default args expansion

The tree created during expansion of default arguments contained trees with the
wrong type of positions. Let's discuss this with an example. Below is the tree
generated for the `foo` method in the test class included in this commit.

Before this commit:

```
[54:94]def foo(): [58]Unit = <70:90>{
  [70:79]<artifact> val qual$1: [70]Bar = [70:79][70:79][70:79]new [74:77]Bar();
  [80]<artifact> val x$1: [80]Int = [80]qual$1.bar$default$1;
  <70:90><70:83>qual$1.bar([80]x$1)
}
```

Now:

```
[54:99]def foo(): [58]Unit = <70:95>{
  <70:84><artifact> val qual$1: [70]Bar = [70:84][70:84][70:84]new [74:77]Bar();
  [85]<artifact> val x$1: [85]Int = [85]qual$1.bar$default$1;
  <70:95>[84:88]qual$1.bar([85]x$1)
}
```

Here are the list of changes:

* The synthetic `qual$1` has a transparent position, instead of a range position.
* The new Select tree (i.e., `qual$1.bar`) should always have a range position,
  because `selected` (i.e., the called method) is always visible in the source
  (in fact, this is the whole point of the fix, we need a range position or
  hyperlinking request from the Scala IDE won't work).
* The Block that contains the expanded default arguments is forced to have a
  transparent position, as it never exist in the original source.

The tricky part of the fix is the position assigned to the new Select tree,
which needs to respect the range position's invariants. In the specific case,
we ought to make sure that range positions don't overlap. Therefore, the position
assigned to the new Select tree is computed by intersecting the original Select
position (i.e., `baseFun`'s position) and the original qualifier's position
(i.e., `qual`'s position).

If you take a closer look at the range positions assigned in the tree after this
commit, you'll notice that the range position of the `qual$1`'s rhs (i.e.,
[70:84]), and `qual$1.bar` (i.e., [84:88]) might seem to overlap, because the
former ends where the latter begins. However, this not the case because of the
range position's invariant 2, which states:

> Invariant 2: in a range position,  start <= point <  end

Hence, the above two positions aren't overlapping as far as the compiler is
concerned.

One additional aspect (that may look like a detail) is that we make sure to
never generate a position such that its start is after its end. This is why we
take the position with the smallest end point.

Furthermore, calling `withStart` would turn any position in a range position,
which isn't desiderable in general (and, even worse, this can lead to
generation of invalid positions - bad offsets - if the computation is performed
on offset positions). Hence, the position's computation is only performed when
both `baseFun` and `qual` positions are range positions. Indeed, I expect this to
be always the case if the compiler is started with -Yrangepos.

dotta commented on src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala:
```diff
@@ -173,13 +173,16 @@ trait NamesDefaults { self: Analyzer =>
           // to still call 'typed' is to correctly infer singleton types, SI-5259.
-          val f = blockTyper.typedOperator(Select(newQual, selected).setSymbol(baseFun1.symbol))
+          val selectPos = 
+            if(qual.pos.isRange && baseFun.pos.isRange) qual.pos.union(baseFun.pos).withStart(Math.min(qual.pos.end + 1 /*increment to skip the dot*/, baseFun.pos.end))
```
I definitely don't like how I do this, do you see a better way? (see my commit message for some information about things are done the way they are)

retronym replied:
"one space for the dot" seems to be too strong an assumption. What about these `Select`s?

```
scala> 1 . toInt
res0: Int = 1

scala> {1}toInt
warning: there were 1 feature warning(s); re-run with -feature for details
res1: Int = 1
```

dotta replied:
My rationale for shitfing the start position was to avoid generating overlapping positions. But, I don't think it matters if the `selectPos` start is indeed after the dot or not (so, yes, my comment is incorrect).

Let's take an example.

```
class Bar {
  def bar(b: Int = 2) {}
}

class Foo {
  def foo() {
    new Bar().bar/*#*/() \\ (1)
    new Bar() . bar/*#*/() \\ (2)
  }
}
```

Tree generated for (1):

```
      <70:90>{
        <70:79><artifact> val qual$1: [70]Bar = [70:79][70:79][70:79]new [74:77]Bar();
        [80]<artifact> val x$1: [80]Int = [80]qual$1.bar$default$1;
        <70:90>[80:83]qual$1.bar([80]x$1)
      };
```

Tree generated for (2):

```
    <102:124>{
        <102:111><artifact> val qual$2: [102]Bar = [102:111][102:111][102:111]new [106:109]Bar();
        [114]<artifact> val x$2: [114]Int = [114]qual$2.bar$default$1;
        <102:124>[112:117]qual$2.bar([114]x$2)
      }
```

Notice that the range position assigned to `qual$1`'s rhs and `qual$1.bar` never overlap.

Now, say I change the above implementation of `selectPos` to the following (no more start offset shifting)

```
val selectPos = if(qual.pos.isRange && baseFun.pos.isRange) qual.pos.union(baseFun.pos).withStart(Math.min(qual.pos.end, baseFun.pos.end))
  else baseFun.pos
```

Here is the resulting tree for case (1):

```
<70:90>{
        <70:79><artifact> val qual$1: [70]Bar = [70:79][70:79][70:79]new [74:77]Bar();
        [80]<artifact> val x$1: [80]Int = [80]qual$1.bar$default$1;
        <70:90>[79:83]qual$1.bar([80]x$1)
      };
```

Now, the range position of the `qual$1` rhs is the same as the start position of `qual$1.bar`.

Am I worrying about a non-issue?

retronym replied:
Did you paste the correct snippet into "Tree generated for (2):"? Looks like it overlaps, and I suspect it is the last example without the overlap-prevention.

> Am I worrying about a non-issue?

Probably not, but did the range position invariant checkers tell you about this problem, or did you notice it independently? If it is the latter, why didn't the invariant checks go off? Should they be updated? Or are they intentionally weakened underneath transparent positions.?

Overall, your explanation makes sense to me.

I'd like to see some of this documentation transferred as you judge best into the commit comment / scaladoc for the new utility method / additional tests.

Almost there!

dotta replied:
> Did you paste the correct snippet into "Tree generated for (2):"? Looks like it overlaps, and I suspect it is the last example without the overlap-prevention.

Right you are. Wrong copy/paste, I've updated the previous comment with the correct data.

> Probably not, but did the range position invariant checkers tell you about this problem, or did you notice it independently? If it is the latter, why didn't the invariant checks go off? Should they be updated? Or are they intentionally weakened underneath transparent positions.?

The range position's invariant checker reports no problem, I'm simply following my intuition of how it should work, which is why I'm not sure I'm doing the right thing.

The implementation of [Positions.overlaps](https://github.com/scala/scala/blob/master/src/reflect/scala/reflect/internal/util/Position.scala#L200) is surely interesting, as it doesn't match my expectations. In fact, according to that definition, two range positions don't overlap if the end of the first matches the start of the second. However, if [Locator(pos)](https://github.com/scala/scala/blob/master/src/reflect/scala/reflect/internal/Positions.scala#L240) is called on a point position such that it is included in two range positions in the same tree, then there is an ambiguity.

> I'd like to see some of this documentation transferred as you judge best into the commit comment / scaladoc for the new utility method / additional tests.

Sure I'll expand my explanation in the commit message once we have an agreement. But, what new utility method? (and, is `new Bar() . bar/*#*/()` the additional test you'd like to see, or something else?)

dotta replied:
Looks like my intuition is indeed wrong! https://github.com/scala/scala/blob/master/src/reflect/scala/reflect/internal/util/Position.scala#L41

So, this is a no problem, hence [this](https://github.com/dotta/scala/commit/24927c880ef672254bfdefda1ee4c09adc03b71c) implementation is correct. So, unless someone disagree, I'll squash the commits and add the additional information in the commit message. \cc @dragos

retronym commented:
Can you please add an example tree to the commit message to provide a concrete example to discuss? Assume the reviewer doesn't remember the ins and outs of the named/default transformation :)

You could then show the diff of compiling that tree before/after this patch, printing with `-Yprint-pos`.

retronym commented:
> I expect this to be always the case if the compiler is started with -Yrangepos

A macro could type check, or expand into, a tree that uses offset positions that requires names/defaults, even in a compiler run using `-Yrange-pos`. So I don't think an assertion is possible.

dotta commented:
@retronym Thanks. Will update the message. By the way, do you have any idea why the kitty is complaining? I don't see how my patch could cause the error:

```
[error] /localhome/jenkins/b/workspace/pr-scala-integrate-partest/scala-partest/src/main/scala/scala/tools/partest/package.scala:203: type Context in package macros is deprecated: Use BlackboxContext or WhiteboxContext instead
[error]   def traceImpl[A: c.WeakTypeTag](c: Context)(a: c.Expr[A]): c.Expr[A] = {
```

retronym commented:
Try rebasing on the lastest master.

adriaanm commented:
PLS REBUILD ALL

Sorry, the partest integration fail is my bad. Fixed by https://github.com/scala/jenkins-scripts/commit/766003ba7fbc94a76971018a282b9cbdd47d6df8. When #3129 was merged, I had to update partest, but forgot to bump the tag of partest against which we integrate.

I've rebuilt the pr-scala-integrate-partest job manually and it succeeded (see https://scala-webapps.epfl.ch/jenkins/job/pr-scala-integrate-partest/)

Should be all green soon!

scala-jenkins commented:
(kitty-note-to-self: ignore 28514095)
:cat: Roger! Rebuilding pr-scala for 7831a051. :rotating_light:

dotta commented:
> Can you please add an example tree to the commit message to provide a concrete example to discuss? Assume the reviewer doesn't remember the ins and outs of the named/default transformation :)

@retronym Done. Let me know if I should add anything more.

retronym commented:
Thanks for the updated comment! I'm really not sure about the `selectPos` / overlapping question. Perhaps you could check it with @dragos?

dragos commented:
It looks like @dotta sorted out the position/overlapping issue. LGTM from my side.

retronym commented:
LGTM, too.

retronym commented:
Actually, could you please squash these two commits before merge?

dotta commented:
> Actually, could you please squash these two commits before merge?

Sure thing!

Only thing I'm missing is an answer to [this comment](https://github.com/scala/scala/pull/3136#discussion_r7685310) (inlined below).

> > I'd like to see some of this documentation transferred as you judge best into the commit comment / scaladoc for the new utility method / additional tests.
> 
> Sure I'll expand my explanation in the commit message once we have an agreement. But, what new utility method? (and, is `new Bar() . bar/*#*/()` the additional test you'd like to see, or something else?)

dotta commented:
I've squashed the commits and updated the commit message with all the information I thought could be helpful. Please, let me know if anything is missing. Otherwise, feel free to :sheep: it :)

JamesIry commented:
ping @retronym

retronym commented:
LGTM, thanks!
